validator/bigquery_table_retention.rego

```diff
@@ -0,0 +1,84 @@
+#
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+package templates.gcp.GCPBigQueryTableRetentionConstraintV1
+
+import data.validator.gcp.lib as lib
+
+deny[{
+	"msg": message,
+	"details": metadata,
+}] {
+	constraint := input.constraint
+	lib.get_constraint_params(constraint, params)
+
+	min_retention_days := lib.get_default(params, "minimum_retention_days", "")
+	max_retention_days := lib.get_default(params, "maximum_retention_days", "")
+
+	asset := input.asset
+	asset.asset_type == "bigquery.googleapis.com/Table"
+
+	# Check if resource is in exempt list
+	exempt_list := params.exemptions
+	matches := {asset.name} & cast_set(exempt_list)
+	count(matches) == 0
+
+	get_diff(asset, min_retention_days, max_retention_days)
+
+	message := sprintf("BigQuery table %v has a retention policy violation.", [asset.name])
+	metadata := {"resource": asset.name}
+}
+
+###########################
+# Rule Utilities
+###########################
+
+# Generate a violation if the resource retention is greater than the maximum number of retention days allowed.
+get_diff(asset, minimum_retention_days, maximum_retention_days) {
+	maximum_retention_days != ""
+	creation_time := to_number(asset.resource.data.creationTime)
+	retention_days_ms := get_ms_of_retention_days(maximum_retention_days)
+	get_expiration_time := object.get(asset.resource.data, "expirationTime", "")
+	get_expiration_time != ""
+	expiration_time := to_number(get_expiration_time)
+
+	diff := expiration_time - creation_time
+	diff > retention_days_ms
+}
+
+# If expirationTime does not exist when looking at the maximum retention, generate a violation.
+get_diff(asset, minimum_retention_days, maximum_retention_days) {
+	maximum_retention_days != ""
+	get_expiration_time := object.get(asset.resource.data, "expirationTime", "")
+	get_expiration_time == ""
```

Comment:
interesting scenario, if a user specifies a maxium, but no minimum, should this cause  a violation?  From the perspective of least surprise, I think that no retention time should be OK here given that the retention time is effectively 0 which is less than the maximum constraint.

Reply:
I think if the user specifies that there should be a maximum retention, then if there is no expiration date, it should generate a violation since that is a resource whose retention is not being tracked / can potentially overshoot the user defined maximum retention.

Reply:
oh, good point, I misunderstood that it defaults to infinity, this seems like the right thing to do then.